### PR TITLE
HDDS-2115 Add acceptance test for createPipeline CLI and datanode list CLI

### PR DESCRIPTION
**Pull request by timmylicheng**
## What changes were proposed in this pull request?
#HDDS-2115 Add acceptance test for createPipeline CLI and datanode list CLI

(Please fill in changes proposed in this fix)

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2115

(Please create an issue in ASF JIRA before opening a pull request,
and you need to set the title of the pull request which starts with
the corresponding JIRA issue number. (e.g. HDDS-XXXX. Fix a typo in YYY.)

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?
Run test-single.sh locally and all passed.

(Please explain how this patch was tested. Ex: unit tests, manual tests)
(If this patch involves UI changes, please attach a screen-shot; otherwise, remove this)

### CODE REVIEW

**Comment by timmylicheng** [comment]
../test-single.sh scm scmcli/datanode.robot
==============================================================================
ozone-datanode :: Smoketest ozone cluster startup
==============================================================================
Run list pipeline                                                     | PASS |
------------------------------------------------------------------------------
ozone-datanode :: Smoketest ozone cluster startup                     | PASS |
1 critical test, 1 passed, 0 failed
1 test total, 1 passed, 0 failed
==============================================================================
Output:  /tmp/smoketest/ozone/result/robot-ozone-ozone-datanode-scm.xml

**Comment by timmylicheng** [comment]
../test-single.sh scm scmcli/pipeline.robot
==============================================================================
ozone-pipeline :: Smoketest ozone cluster startup
==============================================================================
Run list pipeline                                                     | PASS |
------------------------------------------------------------------------------
Run create pipeline                                                   | PASS |
------------------------------------------------------------------------------
ozone-pipeline :: Smoketest ozone cluster startup                     | PASS |
2 critical tests, 2 passed, 0 failed
2 tests total, 2 passed, 0 failed
==============================================================================
Output:  /tmp/smoketest/ozone/result/robot-ozone-ozone-pipeline-scm.xml
Robot framework is not installed, the reports can be generated (sudo pip install robotframework).

File: hadoop-ozone/dist/src/main/smoketest/scmcli/datanode.robot [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+Documentation       Smoketest ozone cluster startup
+Library             OperatingSystem
+Library             BuiltIn
+Resource            ../commonlib.robot
+
+*** Variables ***
+
+
+*** Test Cases ***
+Run list pipeline
```
[the remark this answers is not in the document]
Comment (timmylicheng) [reply hadoop-ozone/dist/src/main/smoketest/scmcli/datanode.robot]: Updated.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/pipeline/CreatePipelineSubcommand.java [hunk 2]
```diff
@@ -60,10 +61,16 @@ public Void call() throws Exception {
...
-      scmClient.createReplicationPipeline(
+      Pipeline pipeline = scmClient.createReplicationPipeline(
           type,
           factor,
           HddsProtos.NodePool.getDefaultInstance());
+
+      if (pipeline != null) {
+        System.out.println(pipeline.getId().toString() +
+            " is created. Factor: " + pipeline.getFactor().toString() +
+            " ,Type: " + pipeline.getType().toString());
```
[the remark this answers is not in the document]
Comment (timmylicheng) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/pipeline/CreatePipelineSubcommand.java]: Updated

File: hadoop-ozone/dist/src/main/smoketest/scmcli/pipeline.robot [hunk 3]
```diff
@@ -25,4 +25,8 @@ Resource            ../commonlib.robot
 *** Test Cases ***
 Run list pipeline
     ${output} =         Execute          ozone scmcli pipeline list
-                        Should contain   ${output}   Type:RATIS, Factor:ONE, State:OPEN
\ No newline at end of file
+                        Should contain   ${output}   Type:RATIS, Factor:ONE, State:
+
+Run create pipeline
+    ${output} =         Execute          ozone scmcli pipeline create
+                        Should contain   ${output}   is created. Factor:
```
[the remark this answers is not in the document]
Comment (timmylicheng) [reply hadoop-ozone/dist/src/main/smoketest/scmcli/pipeline.robot]: Added.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/pipeline/CreatePipelineSubcommand.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (timmylicheng) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/pipeline/CreatePipelineSubcommand.java]: Removed.

**Comment by timmylicheng** [comment]
ERROR: Test execution of /home/runner/work/hadoop-ozone/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.5.0-SNAPSHOT/compose/ozonesecure is FAILED!!!!

Failed acceptance test is irrelevant.

**Comment by hanishakoneru** [comment]
Thanks @timmylicheng for working on this.
The patch LGTM overall. The acceptance test failure you are seeing is because of another issue which was fixed by HDDS-2774. I have rerun the checks again.

**Comment by hanishakoneru** [comment]
@timmylicheng can you please rebase with latest master. We wouldn't get a clean acceptance test run without the fix in HDDS-2774.

**Comment by timmylicheng** [comment]
> @timmylicheng can you please rebase with latest master. We wouldn't get a clean acceptance test run without the fix in HDDS-2774.

Yea sure. We are going merge tip master to HDDS-1564.

**Comment by timmylicheng** [comment]
After merge with master, acceptance fails at s3 gateway test. @hanishakoneru

**Comment by elek** [comment]
> After merge with master, acceptance fails at s3 gateway test.

Can't see the exact error message. The acceptance tests log files are missing from the archive.

Let me try to retrigger it....

**Comment by hanishakoneru** [comment]
I am not able to expand the acceptance test logs. Going to re-trigger the checks again to see if this is a one time problem or persistent.

**Comment by timmylicheng** [comment]
Removing network ozonesecure_default
cp: cannot stat '/home/runner/work/hadoop-ozone/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.5.0-SNAPSHOT/compose/ozonesecure/result/robot-*.xml': No such file or directory
[ ERROR ] Reading XML source '/home/runner/work/hadoop-ozone/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.5.0-SNAPSHOT/compose/result/robot-*.xml' failed: No such file or directory

Try --help for usage information.
cp: cannot stat '/home/runner/work/hadoop-ozone/hadoop-ozone/hadoop-ozone/dev-support/checks/../../../target/acceptance/log.html': No such file or directory
##[error]Process completed with exit code 1.

@hanishakoneru @elek

**Comment by adoroszlai** [comment]
@timmylicheng all acceptance tests failed because SCM could not exit safe mode:

```
2020-01-07T03:46:31.5337614Z SCM is in safe mode.
2020-01-07T03:46:41.4364121Z SCM is in safe mode.
2020-01-07T03:46:50.4562615Z SCM is in safe mode.
2020-01-07T03:47:00.1458169Z SCM is in safe mode.
2020-01-07T03:47:08.0186479Z SCM is in safe mode.
2020-01-07T03:47:17.1032118Z SCM is in safe mode.
2020-01-07T03:47:26.6599830Z SCM is in safe mode.
2020-01-07T03:47:28.6637806Z WARNING! Safemode is still on. Please check the docker-compose files
```

Docker container logs are full of messages like `group-9415212EE472 not found`.  My guess is that this is the same problem as fixed in [HDDS-2679](https://issues.apache.org/jira/browse/HDDS-2679) and [RATIS-783](https://issues.apache.org/jira/browse/RATIS-783).  Can you please try merging from master again?

**Comment by hanishakoneru** [comment]
Thanks @adoroszlai. I am running CI with @timmylicheng 's changes on top of current master to see if rebasing to master would resolve the issue. I ll post the results when it completes.

**Comment by timmylicheng** [comment]
I merged with master and acceptance failed again with a different message. @hanishakoneru @adoroszlai 

hadoop27-mapreduce :: Execute MR jobs                                         
==============================================================================
Execute PI calculation                                                | FAIL |
Test timeout 4 minutes exceeded.
------------------------------------------------------------------------------
Execute WordCount                                                     | FAIL |
255 != 0
------------------------------------------------------------------------------
hadoop27-mapreduce :: Execute MR jobs                                 | FAIL |
2 critical tests, 0 passed, 2 failed
2 tests total, 0 passed, 2 failed
==============================================================================
Output:  /tmp/smoketest/hadoop27/result/robot-hadoop27-hadoop27-mapreduce-rm.xml
/home/runner/work/hadoop-ozone/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.5.0-SNAPSHOT/compose/ozone-mr/hadoop27/../../testlib.sh: line 125: 97441 Killed                  docker-compose -f "$COMPOSE_FILE" --no-ansi logs > "$RESULT_DIR/docker-$OUTPUT_NAME.log"
ERROR: Test execution of /home/runner/work/hadoop-ozone/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.5.0-SNAPSHOT/compose/ozone-mr/hadoop27 is FAILED!!!!
##[error]Process completed with exit code 137.

**Comment by elek** [comment]
Acceptance tests on HDDS-1564 seems to be unstable

**Comment by hanishakoneru** [comment]
Thank you @timmylicheng for the contributions, @adoroszlai and @elek  for the reviews.
+1 to merge into branch HDDS-1564.

**Pull request merged**